File: extract_asar.py
import json
import struct
import os

with open('app.asar', 'rb') as f:
    size_buf = f.read(8)
    header_size = struct.unpack('<I', size_buf[4:8])[0]
    header_json_raw = f.read(header_size)
    header_json = header_json_raw.decode('utf-8').strip('\x00')
    base_offset = 8 + header_size

header = json.loads(header_json)

def extract(dir_node, current_path):
    os.makedirs(current_path, exist_ok=True)
    if 'files' not in dir_node: return
    for name, node in dir_node['files'].items():
        full_path = os.path.join(current_path, name)
        if 'files' in node:
            extract(node, full_path)
        elif node.get('unpacked'):
            print(f"Skipping unpacked: {full_path}")
        else:
            size = int(node['size'])
            offset = int(node['offset'])
            with open('app.asar', 'rb') as f:
                f.seek(base_offset + offset)
                data = f.read(size)
            with open(full_path, 'wb') as out:
                out.write(data)

extract(header, 'app_unpacked')
print("Done!")

File: extract_asar.js
const fs = require('fs');
const path = require('path');

const archivePath = path.resolve('app.asar');
const destPath = path.resolve('app_unpacked');

const fd = fs.openSync(archivePath, 'r');
const buf = Buffer.alloc(16);
fs.readSync(fd, buf, 0, 16, 0);

const strLen = buf.readUInt32LE(12);
const jsonBuf = Buffer.alloc(strLen);
fs.readSync(fd, jsonBuf, 0, strLen, 16);

let jsonStr = jsonBuf.toString('utf8');
// remove trailing nulls padding
jsonStr = jsonStr.replace(/\0+$/, '');

const header = JSON.parse(jsonStr);

// Sometimes offset is relative to the end of the header
// The header size includes the 8 byte prefix usually.
// Wait, the standard electron ASAR offset is relative to sizeBuf.readUInt32LE(4) + 8.
const headerSize = buf.readUInt32LE(4);
const baseOffset = headerSize + 8;

function extract(dirNode, currentPath) {
    if (!fs.existsSync(currentPath)) {
        fs.mkdirSync(currentPath, { recursive: true });
    }
    
    if (!dirNode.files) return;

    for (const name in dirNode.files) {
        const fileNode = dirNode.files[name];
        const fullPath = path.join(currentPath, name);
        
        if (fileNode.files) {
            extract(fileNode, fullPath);
        } else if (fileNode.unpacked) {
            console.log(`Skipping unpacked file: ${fullPath}`);
        } else if (fileNode.size !== undefined && fileNode.offset !== undefined) {
            const offset = parseInt(fileNode.offset);
            const size = parseInt(fileNode.size);
            
            const fileBuf = Buffer.alloc(size);
            fs.readSync(fd, fileBuf, 0, size, baseOffset + offset);
            fs.writeFileSync(fullPath, fileBuf);
        }
    }
}

extract(header, destPath);
fs.closeSync(fd);
console.log('Done extracting!');
